src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -329,6 +342,47 @@ steps Merge {..} requestedSteps = assertStepsInvariant <$> do
             writeMutVar mergeState $! MergingDone
             pure (n + dropped, MergeDone)
 
+{-# SPECIALISE doStepsUnion ::
+     Merge IO h
+  -> Int
+  -> IO (Int, StepResult) #-}
+doStepsUnion ::
+     (MonadMask m, MonadSTM m, MonadST m)
+  => Merge m h
+  -> Int  -- ^ How many input entries to consume (at least)
+  -> m (Int, StepResult)
+doStepsUnion Merge {..} requestedSteps = go 0
+  where
+    go !n
+      | n >= requestedSteps =
+          return (n, MergeInProgress)
+      | otherwise = do
+          (key, entry, hasMore) <- Readers.pop mergeReaders
+          handleEntry (n + 1) key entry hasMore
+
+    -- Similar to 'handleMupdate' in 'stepsLevel', but here we have to combine
+    -- all entries monoidally.
+    handleEntry !n !key !entry Readers.Drained = do
+        -- no future entries, no previous entry to resolve, just write!
+        writeReaderEntry mergeType mergeBuilder key entry
+        writeMutVar mergeState $! MergingDone
+        pure (n, MergeDone)
+
+    handleEntry !n !key !entry Readers.HasMore = do
+        nextKey <- Readers.peekKey mergeReaders
+        if nextKey /= key
+          then do
+            -- resolved all entries for this key, write it
+            writeReaderEntry mergeType mergeBuilder key entry
+            go n
+          else do
+            (_, nextEntry, hasMore) <- Readers.pop mergeReaders
+            -- for resolution, we need the full second value to be present
+            let resolved = combineUnion mergeMappend
+                             (Reader.toFullEntry entry)
+                             (Reader.toFullEntry nextEntry)
+            handleEntry (n + 1) key (Reader.Entry resolved) hasMore
```

Comment:
Note that if 'mergeMappend' is @const@ (e.g. when doing a union on a normal/non-monoidal table), it seems like we potentially do more work than necessary: Once we find the first non-delete entry, we could ignore the remaining entries for that key. However, this depends on union's semantics for blobs. With the current model, if any input table has a blob for a key, the result should also have a blob, so we have to keep looking for blobs.

Maybe that's not the right choice of semantics. If we have multiple blobs for a key, which one should we choose?

Maybe this is also a (rare) example of a situation where the combination of blobs and monoidal resolution (which can't appear with a normal/monoidal table distinction) make an optimisation difficult.

Reply:
> So if I understand correctly there are two questions

Yes, although they are related: if we don't change the semantics for blobs, and don't make the normal/monoidal table distinction explicit in the internals here, then I don't think we gain much from optimising the `const` case.